Added c++ version of the name generator

Kronuz opened the pull request:
Requires C++11

skeeto commented:
This is very interesting, thanks! It looks like you ported the
JavaScript version to C++.

I had trouble building it on my system (Debian Jessie, g++ 4.9),
particularly because I was getting a crazy compiler template error
message. After some investigation, I'm pretty sure Generator::symbols is
ill-defined because vector<const T> isn't permitted in C++. Vector
elements must be MoveAssignable (or does C++11 change this?). Removing
the const fixed the build. I also had to include <cwctype> to get
towupper().

I see the MIT license, but would you consider dedicating your
implementation to the public domain? The other three implementations are
public domain, per the UNLICENSE file in the repository root.

Kronuz commented:
Yes, this is a port from JavaScript (also adds Collapse Triple option). I've updated the code to fix it in gcc (now it works in clang and gcc).

Regarding the license, I've also changed that to Public Domain in the headers.

skeeto commented:
Perfect, thanks!
